spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/ExpireSnapshotsSparkAction.java

```diff
@@ -222,13 +228,34 @@ private ExpireSnapshots.Result doExecute() {
     }
   }
 
+  /**
+   * Builds a dataset of reachable files from given table metadata
+   *
+   * @param metadata table metadata
+   * @return a dataset of files: schema(file_path, file_type)
+   */
   private Dataset<Row> buildValidFileDF(TableMetadata metadata) {
     Table staticTable = newStaticTable(metadata, table.io());
     return buildValidContentFileWithTypeDF(staticTable)
         .union(withFileType(buildManifestFileDF(staticTable), MANIFEST))
         .union(withFileType(buildManifestListDF(staticTable), MANIFEST_LIST));
   }
 
+  /**
+   * Builds a dataset of reachable files from given table metadata, with a snapshot filter
+   *
+   * @param metadata table metadata
+   * @param snapshotsToExclude files reachable by this snapshot will be filtered out
+   * @return a dataset of files: schema(file_path, file_type)
+   */
+  private Dataset<Row> buildFilteredValidDataDF(
+      TableMetadata metadata, Set<Long> snapshotsToExclude) {
+    Table staticTable = newStaticTable(metadata, table.io());
+    return buildValidContentFileWithTypeDF(staticTable, snapshotsToExclude)
```

Comment:
Two questions.

- Will we scan `ALL_MANIFESTS` twice? Once for content files and once for manifests?
- Will we open extra manifests when building the reachability set for expired snapshots? I don't think the current implementation removes any manifests from the expired set that are currently in the table.

Reply:
1. Yes, I think there was a proposal for that by @ajantha-bhat  https://github.com/apache/iceberg/pull/4674
2. Yes, I think I gave it a try but decided to punt for now as it always involves an extra Spark join.  (the all-manifests already being in a dataset).  And so not always beneficial in all cases.

Reply:
> Yes, I think there was a proposal for that by @ajantha-bhat https://github.com/apache/iceberg/pull/4674, we werent sure that caching is worth the cost (it may involves an extra read-write)

I didn't find so much performance difference when tested locally. Hence, I didn't proceed further on #4674. May be I need to Test in S3 to really see the benefit.

Reply:
I was thinking about something like this:

- Cache/persist a dataset of unique manifests in still valid snapshots
- Cache/persist a dataset of unique manifests in expired snapshots
- Find expired manifests that are no longer referenced by still valid snapshots by anti-joining manifest datasets
- Read expired manifests to build a dataset of content files referenced by expired manifests
- Read still valid manifests to build a dataset of still valid content files
- Find expired content files by anti-joining content file datasets
- Find expired manifest lists
- Union datasets for different expired  results

If we do it this way, we read `all_manifests` table only once and also skip still live manifests when looking for expired content files. The extra shuffle is a valid concern but I think we can avoid it. Right now, we are doing a round-robin partitioning when building unique manifests. Instead, we can assign a deterministic hash partitioning to both datasets to avoid a shuffle during anti-joins. Spark won't shuffle datasets if they are partitioned in the same way. We are shuffling that dataset anyway as we deduplicate manifests before opening them up.

@karuppayya is also working on some benchmarks for actions, which would be really handy to evaluate the impact here.

Reply:
Thoughts, @szehon-ho @RussellSpitzer @rdblue?

Reply:
I've kinda been thinking we probably don't need the anti-joins distributed (mostly we just want distributed IO) for this operation but I think all of your logic makes sense here. if I think this through we have

ALL_MANIFESTS broken into 3 subsets
oldManifests: [Manifests only used in Expired Snapshots ] 
commonManifests: [ Manifests used in both expired and unexpired Snapshots ] 
newManifests: [ Manifests only used in unexpired-snapshots ] 

To build this we read allManifests
expiredManifests = allManifests where snapshot is in snapshots being expired
unexpiredManifests = allManifests where snapshots is not in snapshots being expired

oldManifests = expiredManifests -- unexpiredManifests
newManifests = unexpiredManifests -- expiredManifests

We read all the mainfests in oldManifests and newManifests
oldManifests ->  flatmap read manifest and dedupe -> oldFiles
newManifests ->  flatmap read manifest and dedupe -> newFiles

filesToDelete == (oldFiles -- newFiles) + oldManifests + oldManifestLists

Did I get that right?